django/db/models/fields/files.py

```diff
@@ -91,7 +91,14 @@ def open(self, mode="rb"):
     def save(self, name, content, save=True):
         name = self.field.generate_filename(self.instance, name)
         self.name = self.storage.save(name, content, max_length=self.field.max_length)
-        setattr(self.instance, self.field.attname, self.name)
+
+        # The presumption is that IO for content has already been dealt with.
+        setattr(self.instance, self.field.attname, content)
+
+        # Update the name in case generate_filename or the storage backend
+        # changed the name. Avoids re-running the descriptor logic.
+        self.instance.__dict__[self.field.attname] = self.name
```

Comment:
This appears to be the core of the fix and all other changes in the file can be reverted and the test still passes. 🤔 
Either we need more tests to show the refactoring is needed or we should revert the other changes

Reply:
Lol. Maybe.  I only arrived at that after rewriting about a five times though.

Reply:
Yes! John you've put loads of work into this and helped improve the understanding of an area not been touched in years ⭐️
Changes that end up being a few lines are never easy and take hours and hours to come to. Thank you for the work you've invested in this